skops/card/_model_card.py

```diff
@@ -202,7 +275,51 @@ def add_plot(self, **kwargs: str) -> "Card":
             Card object.
         """
         for plot_name, plot_path in kwargs.items():
-            self._extra_sections[plot_name] = plot_path
+            section = PlotSection(alt_text=plot_name, path=plot_path)
+            self._extra_sections[plot_name] = section
+        return self
+
+    def add_table(self, **kwargs: dict["str", list[Any]]) -> Card:
+        """Add a table to the model card.
+
+        Add a table to the model card. This can be especially useful when you
+        using cross validation with sklearn. E.g. you can directly pass the
+        result from calling ``cross_validate`` or the ``cv_results_`` attribute
+        from any of the hyperparameter searches, such as ``GridSearchCV``.
+
+        Morevoer, you can pass any pandas ``DataFrame`` to this method and it
+        will be rendered in the model card. You may consider selecting only a
+        part of the table if it's too big:
+
+        .. code:: python
+
+            search = GridSearchCV(...)
+            search.fit(X, y)
+            df = pd.DataFrame(search.cv_results_)
+            # show only top 10 highest scores
+            df = df.sort_values(["mean_test_score"], ascending=False).head(10)
+            model_card = skops.card.Card(...)
+            model_card.add_table(**{"Hyperparameter search results top 10": df})
+
+        Parameters
+        ----------
+        **kwargs : dict
+            The keys should be strings, which will be used as the section
+            headers, and the values should be tables. Tables can be either dicts
+            with the key being strings that represent the column name, and the
+            values being lists that represent the entries for each row.
+            Alternatively, the table can be a pandas ``DataFrame``. The table
+            must not be empty.
+
+        Returns
+        -------
+        self : object
+            Card object.
+
+        """
+        for key, val in kwargs.items():
+            section = TableSection(table=val)
+            self._extra_sections[key] = section
```

Comment:
here we could add to `self._extra_tables` instead and then we know they're tables. Each method adds to their corresponding attribute.

Reply:
I wanted to avoid having separate containers for each type of addition. One reason is that it makes the code more lengthy (we'll later have to loop over each one separately), the other is that we lose the information of which order the extra sections are stored (e.g. I add a plot, a table, then a plot, but at the end I get plot, plot, table).

Another design I wanted to avoid is formatting already inside the `add_*` method, to not lock us in into a specific format at that point.

Reply:
okay, so if the order is important (which I agree with), we should make it a list of named tuples instead of a dict (and relying on the dict preserving insertion order during iteration).

If we remove the protocol and not typing our private attribute, then I'm fine with the PR

Reply:
Happy to make those changes.

Regarding the dict to list of tuples proposal: From my understanding, as long as you only add to the dict, the order should be preserved, right? So the difference would mainly be if we want to later allow deletion, or if a user wants to have multiple sections with the same name?

Reply:
that's true in modern new python versions. I think of it as an implementation detail, and rather not rely on it really.

Reply:
When the dict change was made in Python 3.6, it was indeed so:

> The order-preserving aspect of this new implementation is considered an implementation detail and should not be relied upon

([link](https://docs.python.org/3/whatsnew/3.6.html#new-dict-implementation))

With Python 3.7 (which is our min version), the order is officially part of the language:

> the insertion-order preservation nature of [dict](https://docs.python.org/3/library/stdtypes.html#typesmapping) objects [has been declared](https://mail.python.org/pipermail/python-dev/2017-December/151283.html) to be an official part of the Python language spec.

([link](https://docs.python.org/3/whatsnew/3.7.html))

So this wouldn't be my main concern here. I'd rather think about these implications:

1. We can have duplicate section names in the extra sections
2. Time complexity to retrieve items is now O(N) and there is no guarantee that only a single key matches (due to 1.)
3. More control about order (e.g. when deleting and reinserting items)

Due to the nature of the extra sections, which can contain anything, I think the trade-off is okay, whereas template sections should remain a dict.

Reply:
I know all these details about python's dict, but not many people do, and a "dictionary" isn't necessarily ordered in other languages. I tend to avoid things which are somewhat common in different languages but slightly differ in python for some specific reason.